### PR TITLE
Add actionable item to PatternMatchExhaustivity diagnostic

### PR DESCRIPTION
**Pull request by ghostbuster91**
The purpose of this PR is to add an actionable item for non-exhaustive pattern match diagnostic, so that people can auto insert missing cases. 

Relates to https://github.com/scalameta/metals-feature-requests/issues/350

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala
Comment (ghostbuster91) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:898]: The `Space` type was probably never meant to be used to render the actual scala code. I wonder if there is a sane way to accomplish that.

File: compiler/test/dotty/tools/dotc/reporting/CodeActionTest.scala
Comment (ghostbuster91) [inline compiler/test/dotty/tools/dotc/reporting/CodeActionTest.scala:97]: This test fails because string literals are rendered without surrounding quotes.

**Comment by ghostbuster91** [comment]
@ckipp01 could you give it a look?

File: compiler/test/dotty/tools/dotc/reporting/TestReporter.scala
Comment (dwijnand) [inline compiler/test/dotty/tools/dotc/reporting/TestReporter.scala:78]:
```suggestion
      case pm: PatternMatchExhaustivity => s": ${pm.uncovered.mkString(", ")}"
```

File: compiler/src/dotty/tools/dotc/reporting/messages.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/reporting/messages.scala:846]: You're changing the message from being the first 6 examples, comma-separated, to all examples rendered as a list (which is bracketed).

**Review by dwijnand — CHANGES_REQUESTED** [review]

File: compiler/test/dotty/tools/dotc/reporting/TestReporter.scala
Comment (ghostbuster91) [reply compiler/test/dotty/tools/dotc/reporting/TestReporter.scala]: I fixed it by adding an `uncovered` property that contains already preformatted message.

File: compiler/src/dotty/tools/dotc/reporting/messages.scala
Comment (ghostbuster91) [reply compiler/src/dotty/tools/dotc/reporting/messages.scala:846]: Good catch, I forgot to restore `take(6)`. Fixed together with list rendering.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala
Comment (rochala) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:788]: I believe it can be replaced with:
```scala
case Typ(c: ConstantType, _) => c.show
```

**Comment by ckipp01** [comment]
Just a note about something we need to figure out. Testing this I now see our e2e with Dotty/sbt/Metals isn't working as expected. I know there was some recent changes in sbt, but the behavior I'm seeing now is that the diagnostic is published with the actions, but then published again without them with `reset` to `true`. For example:

```
[Trace - 00:47:25 pm] Received notification 'build/publishDiagnostics'
Params: {
  "textDocument": {
    "uri": "file:///Users/ckipp/Documents/scala-workspace/actionable-diagnostic/src/main/scala/example/Hello.scala"
  },
  "buildTarget": {
    "uri": "file:/Users/ckipp/Documents/scala-workspace/actionable-diagnostic/#root/Compile"
  },
  "diagnostics": [
    {
      "range": {
        "start": {
          "line": 17,
          "character": 26
        },
        "end": {
          "line": 17,
          "character": 30
        }
      },
      "severity": 2,
      "code": "29",
      "source": "sbt",
      "message": "\u001b[33mmatch\u001b[0m may not be exhaustive.\n\nIt would fail on pattern case: Tree.Leaf(_)\n",
      "relatedInformation": [],
      "dataKind": "scala",
      "data": {
        "actions": [
          {
            "title": "Insert missing cases (1)",
            "edit": {
              "changes": [
                {
                  "range": {
                    "start": {
                      "line": 18,
                      "character": 33
                    },
                    "end": {
                      "line": 18,
                      "character": 33
                    }
                  },
                  "newText": "\n      case Tree.Leaf(_) \u003d\u003e ???"
                }
              ]
            }
          }
        ]
      }
    }
  ],
  "reset": false
}


[Trace - 00:47:25 pm] Received notification 'build/logMessage'
Params: {
  "type": 2,
  "message": "\u001b[33m\u001b[33m-- [E029] Pattern Match Exhaustivity Warning: /Users/ckipp/Documents/scala-workspace/actionable-diagnostic/src/main/scala/example/Hello.scala:18:26 \u001b[0m\u001b[0m\n\u001b[33m18 |\u001b[0m    \u001b[33mdef\u001b[0m \u001b[36mfoo\u001b[0m(\u001b[36mtree\u001b[0m: \u001b[35mTree\u001b[0m) \u003d tree \u001b[33mmatch\u001b[0m {\n\u001b[33m\u001b[33m   |\u001b[0m                          ^^^^\u001b[0m\n\u001b[33m   |\u001b[0m                          \u001b[33mmatch\u001b[0m may not be exhaustive.\n\u001b[33m   |\u001b[0m\n\u001b[33m   |\u001b[0m                          It would fail on pattern case: Tree.Leaf(_)\n\u001b[33m   |\u001b[0m\n\u001b[33m   |\u001b[0m longer explanation available when compiling with `-explain`"
}


[Trace - 00:47:25 pm] Received notification 'build/logMessage'
Params: {
  "type": 2,
  "message": "one warning found"
}


[Trace - 00:47:25 pm] Received notification 'build/logMessage'
Params: {
  "type": 3,
  "message": "done compiling"
}


[Trace - 00:47:25 pm] Received notification 'build/publishDiagnostics'
Params: {
  "textDocument": {
    "uri": "file:///Users/ckipp/Documents/scala-workspace/actionable-diagnostic/src/main/scala/example/Hello.scala"
  },
  "buildTarget": {
    "uri": "file:/Users/ckipp/Documents/scala-workspace/actionable-diagnostic/#root/Compile"
  },
  "diagnostics": [
    {
      "range": {
        "start": {
          "line": 17,
          "character": 26
        },
        "end": {
          "line": 17,
          "character": 30
        }
      },
      "severity": 2,
      "source": "sbt",
      "message": "\u001b[33mmatch\u001b[0m may not be exhaustive.\n\nIt would fail on pattern case: Tree.Leaf(_)\n",
      "relatedInformation": []
    }
  ],
  "reset": true
}
```

I'm not really sure why this is happening now.

**Comment by tgodzik** [comment]
I think this was already fixed in https://github.com/sbt/zinc/issues/1225 ?

We had same issues when discussing https://github.com/sbt/zinc/discussions/1214

**Comment by ckipp01** [comment]
> I think this was already fixed in [sbt/zinc#1225](https://github.com/sbt/zinc/issues/1225) ?
> 
> We had same issues when discussing [sbt/zinc#1214](https://github.com/sbt/zinc/discussions/1214)

Yea, it's just odd that we didn't hit on this before with the other examples for Scala 3. I'm trying to update https://github.com/lampepfl/dotty/blob/10503b0345550259e9381f4fe59da80fb35b32b9/sbt-bridge/src/dotty/tools/xsbt/CompilerBridgeDriver.java#L139-L140 right now which I also believe might be problematic here because it's not using the new `problem2`.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala
Comment (ghostbuster91) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Almost, but it put me on the right track, thanks!

**Comment by ckipp01** [comment]
Alright I hope it's ok, but I just rebase this @ghostbuster91 because I want to test it now that https://github.com/lampepfl/dotty/pull/18322 is merged in.

**Comment by ghostbuster91** [comment]
> Alright I hope it's ok, but I just rebase this @ghostbuster91 because I want to test it now that https://github.com/lampepfl/dotty/pull/18322 is merged in.

Nice, thanks:) Could you also share how you are testing it together with these e2e tests with metals?

**Comment by ckipp01** [comment]
> Nice, thanks:) Could you also share how you are testing it together with these e2e tests with metals?

Sure, I know you're using `nvim-metals` as well, so the steps should be the same:

1. Do a `sbt publishLocal` on this repo
2. Create a minimal sbt project using 1.9.3
3. Set the `scalaVersion` to the version you just published locally
4. Switch your build server to use `sbt` instead of Bloop
5. Then just go into your main file and drop the code from your test in there. You should see a diagnostic and be able to trigger a code action on it.
6. If you don't you can start debugging by creating a `.metals/bsp.trace.json` file and look in the publishDiagnostics and see what they look like, _or if they are there at all_.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala
Comment (ckipp01) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:904]: To be honest I'm not really sure.

**Review by ckipp01 — CHANGES_REQUESTED** [review]
This looks great @ghostbuster91! Thanks for doing this! Just one small nit on the test, let's make sure we also have one with significant whitespace being used just to be safe.

File: compiler/src/dotty/tools/dotc/reporting/messages.scala
Comment (ghostbuster91) [reply compiler/src/dotty/tools/dotc/reporting/messages.scala:875]: Good idea, added.

**Comment by ghostbuster91** [comment]
> This looks great @ghostbuster91! Thanks for doing this! Just one small nit on the test, let's make sure we also have one with significant whitespace being used just to be safe.

Thanks for making it possible :)

**Review by ckipp01 — APPROVED** [review]
This looks good to me. Thanks a lot @ghostbuster91! LGTM

**Comment by ckipp01** [comment]
It's not actually letting me merge. @dwijnand I think since you requested changes it's waiting for your approval.

**Review by dwijnand — APPROVED** [review]

**Pull request merged**
